### PR TITLE
[TE] MaxText permute padding and alignment

### PR DESCRIPTION
**Pull request by jberchtold-nvidia**
# Description

Add support for padding and aligning groups to MaxText's pure-JAX permute codepath. In local testing, this is faster than  the current TE router+permute kernels, mostly due to the TE permute kernels.

This alignment and padding is required for the TE grouped GEMM.

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -294,7 +294,8 @@ def __init__(
       quant_dot_general = nnx_wrappers.ToNNX(dot_general_linen, rngs=rngs)
       self._quant_dot_general_name = f"{type(dot_general_linen).__name__}_0"
       setattr(self, self._quant_dot_general_name, quant_dot_general)
-      dummy_inputs = jnp.zeros((1, *self.in_features_shape), dtype=self.dtype)
+      block_size = getattr(quant, "get_block_size", lambda: 1)()  # needed for TE block scaling
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py:297]: Ignore this change, this is unrelated and from this PR: #35

File: src/maxtext/layers/moe.py [hunk 2]
```diff
@@ -698,6 +699,34 @@ def _mt_permute(self, inputs, gate_logits, pre_bias_logits, use_custom_sort_vjp=
 
     weights, selected_experts = self.get_topk(gate_logits, pre_bias_logits, rngs)
 
+    token_count_per_expert = jnp.bincount(selected_experts.ravel(), length=self.num_experts)
```
Comment (tdophung) [inline src/maxtext/layers/moe.py]: Nit SR: use `tokens_per_expert` instead so it is consistent with other places (TE files)

File: src/maxtext/layers/moe.py [hunk 3]
```diff
@@ -698,6 +699,34 @@ def _mt_permute(self, inputs, gate_logits, pre_bias_logits, use_custom_sort_vjp=
...
+    align_size = self.config.moe_permutation_group_align_size
+    if align_size > 0:
+      # Use (align - count % align) % align so that experts already aligned to
```
Comment (tdophung) [inline src/maxtext/layers/moe.py]: `tokens_per_experts`

File: src/maxtext/layers/moe.py [hunk 4]
```diff
@@ -698,6 +699,34 @@ def _mt_permute(self, inputs, gate_logits, pre_bias_logits, use_custom_sort_vjp=
...
+      # Each expert i gets a slot of (align_size - 1) positions (worst-case padding,
+      # which occurs when token_count[i] % align_size == 1).
+      # Within slot i: positions where offset < padding_tokens_required_per_expert[i]
+      # are assigned to expert i (real padding); the rest point to the last expert
+      # as overflow placeholders to keep the buffer statically sized.
+      # E.g. padding_tokens_required_per_expert=(1,3), align_size=8 → total=14:
+      #   [0, last, last, last, last, last, last,   # slot 0: 1 real, 6 overflow
+      #    1, 1, 1, last, last, last, last]          # slot 1: 3 real, 4 overflow
+      max_padding_per_expert = align_size - 1
+      total_padding_size = self.num_experts * max_padding_per_expert
```
Comment (tdophung) [inline src/maxtext/layers/moe.py]: Nit SR: `max_total_padding_size`

File: src/maxtext/layers/moe.py [hunk 5]
```diff
@@ -698,6 +699,34 @@ def _mt_permute(self, inputs, gate_logits, pre_bias_logits, use_custom_sort_vjp=
...
+    token_count_per_expert = jnp.bincount(selected_experts.ravel(), length=self.num_experts)
+    align_size = self.config.moe_permutation_group_align_size
+    if align_size > 0:
+      # Use (align - count % align) % align so that experts already aligned to
+      # align_size need 0 padding tokens (not align_size, which would exceed the
+      # per-expert slot capacity of align_size - 1).
+      padding_tokens_required_per_expert = (align_size - (token_count_per_expert % align_size)) % align_size
+      # Build a static-size padding index buffer.
+      # Each expert i gets a slot of (align_size - 1) positions (worst-case padding,
+      # which occurs when token_count[i] % align_size == 1).
```
Comment (tdophung) [inline src/maxtext/layers/moe.py]: Nit SR: tokens_per_experts[e]

File: src/maxtext/layers/quantizations.py [hunk 6]
```diff
@@ -948,14 +949,27 @@ def einsum(self, *args, **kwargs):
     # quant.einsum is only required for MoE or for inference with KVCache.
     raise ValueError("Einsum is not yet supported for TransformerEngine quantization.")
 
-  def gmm(self, inputs, kernel, tiling, group_sizes, expert_assignments):
+  def get_gmm_align_size(self, te_gmm_quantization_recipe_name: str):
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/quantizations.py:952]: These changes are from this PR: #38

File: src/maxtext/layers/moe.py [hunk 7]
```diff
@@ -1524,7 +1581,7 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
         # The ring-of-experts strategy first duplicates the inputs to all
         # expert shards, and then routes within each shard.
 
-        use_te = self.config.te_router_and_permutation_impl and te_permutation.TE_PERMUTATION_AVAILABLE and te_router.TE_ROUTER_AVAILABLE
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py:1527]: Removing silent fallback I missed in previous PR. In practice, this may not actually do a silent fallback as earlier guards will ensure TE permute is available but keeping this consistent with other checks

File: src/maxtext/layers/moe.py [hunk 8]
```diff
@@ -1549,9 +1606,10 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
         )
 
         # Filter down to the group sizes that apply to only the experts in the
-        # current shard. ragged_dot only processes sum(group_sizes) tokens,
-        # so no input masking is needed.
+        # current shard.
         group_sizes = group_sizes[:num_experts_per_shard]
+        mask = jnp.arange(x.shape[0]) < jnp.sum(group_sizes)
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py]: For the MT permute+padding, I needed to revert this change that previously removed this in #36, otherwise I got NaN in loss. I need to verify why the MT permute needs this and TE permute doesn't and gate this behind MT permute only

File: src/maxtext/layers/moe.py [hunk 9]
```diff
@@ -1549,9 +1606,11 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
         )
 
         # Filter down to the group sizes that apply to only the experts in the
-        # current shard. ragged_dot only processes sum(group_sizes) tokens,
-        # so no input masking is needed.
+        # current shard.
         group_sizes = group_sizes[:num_experts_per_shard]
+        if not use_te:
+          mask = jnp.arange(x.shape[0]) < jnp.sum(group_sizes)
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py:1612]: Adding back in this masking for the MT permute codepath

**Review by tdophung — APPROVED** [review]
LGTM
